### PR TITLE
Enabling Robottelo for IPv6 Testing

### PR DESCRIPTION
**Pull request by jyejare**
### Problem Statement
Enabling IPv6 switching in Robottelo Automation 

### Solution
- Added the IPV6 switch and IPV6 proxy config that would switch ON and OFF the IPv6 Testing for robottelo tests via CI or local
- Added setup of Satellite HTTP Proxy by default to all satellite spinned as/by broker workers, factory, directly from Broker context manager
- Adder teardown of Satellite HTTP Proxy by default to all satellite spinned as/by broker workers, factory, directly from Broker context manager
- Added setup of Capsule HTTP Proxy by default to all capsules spinned by factory, directly from Broker context manager to speak to CDN for content access
- Adder teardown of Capsule HTTP Proxy by default to all capsules spinned by factory, directly from Broker context manager
- Deploy Arguments switches satellite / capsule spin up to use Ipv6 sat/cap/host

**Followup PRs, coming next:**
- [DEVs are yet to decide, Out of scope of the PR] Hosts spins up to use ipv6/ipv4 network (based on strategy of host testing in IPv6 is yet to decide)
- HTTP Proxy specific tests uses its own non-default-http-proxy setup 

### Broker Pending(Not blocking this PR from merge):
- Adding support to deploy container as content host in Ipv6 network

### CODE REVIEW

**Review by JacobCallahan — CHANGES_REQUESTED** [review]
can you add some validators for these?

File: robottelo/hosts.py
Comment (ogajduse) [inline robottelo/hosts.py:1822]: Is there a scenario when a Satellite can have more than one HTTP proxy? If not, could we have the proxy as a Satellite class parameter? We would not need to specify what HTTP proxy to disable.

File: robottelo/hosts.py
Comment (ogajduse) [inline robottelo/hosts.py:1]: More than 90% of the functions and methods in `hosts.py` use an API interface. Is there any particular reason why you used the CLI interface to manipulate with HTTP proxy and Settings entities?

**Review by ogajduse — CHANGES_REQUESTED** [review]

File: robottelo/hosts.py
Comment (jyejare) [reply robottelo/hosts.py:1792]: @ogajduse Yeah, you are right, its not yet because we are still waiting for IPv6 playground in CI to run test execution for IPV6.

I opened this PR for review for these static checks, so that we curb the issues earlier before we run it actually in CI.

Comment (jyejare) [reply robottelo/hosts.py:1792]: BTW, the all suggestion is modified now to:
```
all([settings.server.is_ipv6, settings.server.http_proxy_ipv6_url])
```

File: robottelo/hosts.py
Comment (jyejare) [reply robottelo/hosts.py:1815]: The `if not` find proxy creates a new proxy and `else` condition searches the proxy(which we already searched it exists in `if not` condition). In both conditions it returns the proxy object.

So since anyways we are creating / reading the existing proxy, we dont need to raise an exception.

File: pytest_fixtures/core/sat_cap_factory.py
Comment (jyejare) [reply pytest_fixtures/core/sat_cap_factory.py:336]: I would hold off the answer a bit because, I am interested to know about whether Teams really want to teardown it since we have components split CI jobs. 

Also, the disable is written to disable based on the needs per test/module/component for now, I just added it everywhere to show Teams how it looks like when implemented.

File: pytest_fixtures/core/sat_cap_factory.py
Comment (jyejare) [reply pytest_fixtures/core/sat_cap_factory.py:384]: Good catch, it should be before checkin.

File: robottelo/hosts.py
Comment (jyejare) [reply robottelo/hosts.py:1822]: Agree. But I am not sure if we could have more than one or not. I would like Teams to respond on that request.

Ping @vsedmik

File: robottelo/hosts.py
Comment (jyejare) [reply robottelo/hosts.py:1]: No special reason! I can move on that to use API.

File: robottelo/hosts.py
Comment (vsedmik) [reply robottelo/hosts.py:1818]: In Satellite you can have defined more HTTP proxy entities. But for communication you can use just one. I mean one for Foreman (http_proxy) and one for Katello (content_default_http_proxy). I believe we need to set both of them in settings to route all traffic from Satellite through them.

Comment (jyejare) [reply robottelo/hosts.py:1818]: @vsedmik So can we create one http proxy (as we are doing here) and set it for both foreman and katello? Do we have different settings for both of them which we can set ?

Comment (vsedmik) [reply robottelo/hosts.py:1818]: Correct, create one HTTP proxy and set it Settings in `http_proxy` and `content_default_http_proxy`. I'm not aware of any other similar settings we should set.

Comment (jyejare) [reply robottelo/hosts.py:1818]: Thanks @vsedmik , I can fix that right away.

Comment (jyejare) [reply robottelo/hosts.py:1825]: Fixed!

File: pytest_fixtures/core/sat_cap_factory.py
Comment (jyejare) [reply pytest_fixtures/core/sat_cap_factory.py:350]: @ogajduse I am not going to hold it anymore for discussion with teams as its not really useful , I recall that its us / JPL who should decide it. So I would like to continue the discussion on this comment.

File: conf/migrations.py
Comment (rplevka) [inline conf/migrations.py:70]: do I understand it correctly, that this turns everything into `ipv4` or `ipv6` based on a single setting?
aren't we effectively cutting ourselves off from eventual support of mixed stack in the future (e.g. ipv6 satellite test with some ipv4 resource)?

Comment (rplevka) [reply conf/migrations.py:70]: I also believe that config migrations have been introduced to serve different purpose than this.

Another issue i am having with this approach is that you are hardcoding our internal infra design decision into the project (it was our infra decision to keep same hostnames with the changing `ipv` component in its name)

**Review by rplevka — COMMENTED** [review]
I am not 100% sold on the implemented approach to dynamically update the hostname based on a single setting.

My suggestion would be to extend the settings that point to certain hostnames, to contain both versions, so the fixtures/tests could pick the appropriate one.

```
SETTING:
    SOME_SERVER_RESOURCE:
        IPV4: foo.bar.ipv4.baz
        IPV6: foo.bar.ipv6.baz
```
we could also implement some "default" flag (like we do in `broker` with providers) to save ourselves from having to be explicit about the IP version in scenarios where we don't care.

**Comment by jyejare** [comment]
@rplevka The hook implementation isnt really helping here as its resetting all the settings after the `post` hook and hence I changed the implementation to do this switching in config after settings object is initialized.

Coming back to your first concern which is having two different settings is not idealistic because then you have lot of handling throughout robottelo to choose between v4 or v6 instance based on the what network based execution we are doing. Trust me it would be lot of code changes.

With the current approach, the setting names would remain same and only the values/hostnames would change behind the scene, even one doesn't need to change the hostname to switch.

The only switch is `is_ipv6` setting.

File: pytest_fixtures/core/sat_cap_factory.py
Comment (jyejare) [reply pytest_fixtures/core/sat_cap_factory.py:350]: @ogajduse Any further thohghts before I resolve the comment as is ?

Comment (JacobCallahan) [reply pytest_fixtures/core/sat_cap_factory.py:350]: I'm also interested in moving this to setup/teardown methods for the Host class to reduce the need for repetition here.

File: robottelo/hosts.py
Comment (JacobCallahan) [inline robottelo/hosts.py:1851]:
```suggestion
        """Execute procedures for disabling IPv6 HTTP Proxy"""
```

File: robottelo/hosts.py
Comment (JacobCallahan) [inline robottelo/hosts.py:1815]:
```suggestion
        """Execute procedures for enabling IPv6 HTTP Proxy"""
```

File: robottelo/hosts.py
Comment (JacobCallahan) [inline robottelo/hosts.py:1846]:
```suggestion
```

File: robottelo/utils/url.py
Comment (JacobCallahan) [inline robottelo/utils/url.py:17]:
```suggestion
    if isinstance(text, str) and '.ipv4.' in text:
```

**Review by JacobCallahan — CHANGES_REQUESTED** [review]
Just a few minor things

File: robottelo/utils/url.py
Comment (jyejare) [reply robottelo/utils/url.py:17]: @JacobCallahan There is no dots around ipv4. Mostly they are `-ipv4`. That is hostname appended with `-ipv4`

File: robottelo/hosts.py
Comment (jyejare) [reply robottelo/hosts.py:1857]: @JacobCallahan pre-commit complaints about non-None return.

**Review by ogajduse — CHANGES_REQUESTED** [review]
I reviewed changes in all files except `conf/dynaconf_hooks.py` and `robottelo/utils/url.py`. I need to check the branch out on my local machine and play with it a bit. I will do a second review round tomorrow.

File: pytest_fixtures/core/sat_cap_factory.py
Comment (jyejare) [reply pytest_fixtures/core/sat_cap_factory.py:350]: As discussed yesterday in 1:1, having a teardown is not necessary as once the test execution is done on the satellite (be its xdist sat or destructive sat), the satellite will be either checked-in OR even if it kept for debugging there is no harm in keeping that satellite proxy enabled.

There are few components could be worried about keeping pre-enabled proxy and that is `HTTP Proxy` component itself and which has its own custom HTTP proxy servers (auth, unauth etc). That component can disable the default HTTP-proxy during the component setup phase or checkout its own on-demand sat. I would leave that up to the component team to decide how do they want to tackle it.

Overall, I feel like there is no need of explicit teardown here unless asked explicitly.

File: robottelo/utils/url.py
Comment (jyejare) [reply robottelo/utils/url.py]: @ogajduse You mean, you are in favor of keeping `redhat.com`.

Comment (jyejare) [reply robottelo/utils/url.py]: I have moved the condition to check `if '-ipv4' in text`.

File: conf/server.yaml.template
Comment (jyejare) [reply conf/server.yaml.template:17]: @ogajduse Good question. but there would be cross testing in future as well like :
- ipv6 sat with ipv4 cap
- ipv6 sat with ipv4 host

I haven't thought about that design yet though since we are yet to plot that matrix. So I would keep this as is for now because all Ipv6 testing for now is being derived from server network type. We can always change the settings in future as design changes .

File: robottelo/utils/ohsnap.py
Comment (jyejare) [reply robottelo/utils/ohsnap.py:1]: You are right, but I didnot want to wait for that to be done, hence I have added this as a workaround to proceed further. Anyways testing ohsnap connectivity is not part of satellite testing so no matter what we have used for communication.

File: robottelo/hosts.py
Comment (jyejare) [reply robottelo/hosts.py:888]: I would leave it for the future / teams as we are yet to do extensive testing by Teams and they can follow the route we have set in this PR later.

File: robottelo/host_helpers/contenthost_mixins.py
Comment (jyejare) [reply robottelo/host_helpers/contenthost_mixins.py:102]: Its already there used in `_dogfood_helper` function to not surprise :)

The only way we have is to override the function in Satellite class, because though its defined in ContentHost class its used for Satellite in this case.

Comment (jyejare) [reply robottelo/host_helpers/contenthost_mixins.py:102]: Also there are many places in Host classes which are not adhering to the class in which context they are. The best way to clean is to override the method in child class and call super method.

For this PR I don't want to sign up for that work, we could see it as a separate effort.

**Comment by jyejare** [comment]
@ogajduse I have addressed/responded to your comments.  BTW This PR is not to full proof the Ipv6 enablement for all cases. I must say that its the foundation. This PR mostly focuses on Satellite server on Ipv6 network.

We will have many combinations going fwd and the definitions here will change going fwd since we haven't yet decided the combinations yet. So I recommend to review this PR as stepping stone and not full proof solution. 

Thanks for understanding.

CC: @JacobCallahan

File: robottelo/hosts.py
Comment (JacobCallahan) [reply robottelo/hosts.py]: @jyejare I'm not sure I understand what you mean. Can you rephrase your response?

**Comment by jyejare** [comment]
trigger: test-robottelo
pytest: tests/foreman/ -m build_sanity

**Comment by jyejare** [comment]
trigger: test-robottelo
pytest: tests/foreman/ -m build_sanity

**Comment by jyejare** [comment]
The test collection is failing with error:
```"'DynaBox' object has no attribute 'is_ipv6'"```

So the testing has to be run from CI using MR and this PR which has been done there. I shall post results here.

**Review by ogajduse — COMMENTED** [review]
As we discussed on a call today, this is a last portion of comments from my side. I also have some objections about the recursive logic that goes through the settings. I will send more details in a DM to Jitendra since I do not have the suggestion fully shaped.

I will leave the rest of the work on this PR on you, Jake and Jitendra.

File: robottelo/utils/url.py
Comment (jyejare) [reply robottelo/utils/url.py:25]: We dont need to explicitly do that since all other blocks are `elif` and there is nothing to execute after conditions with loop.

File: robottelo/utils/url.py
Comment (jyejare) [reply robottelo/utils/url.py:21]: Solution 2 is already in place and to do that we changing it to dot connected settings. I would rather change the setting to be named string(for unifirmality) instead of number or change the interger to string. 

@Griffin-Sullivan thoughts ?

File: robottelo/hosts.py
Comment (jyejare) [reply robottelo/hosts.py:1784]: The question is out of scope of this PR. But I think it is because the modularity started in RHEL8 but the condition could be changed to `self.os_version.major == 8`, I shall leave it to teams to fix.

**Review by JacobCallahan — APPROVED** [review]
ACK. Let's get this in soon, so we can start really testing it and seeing where pain points show up.

**Review by rmynar — APPROVED** [review]
ACK - although it's not perfect yet, the main goal seems to be completed - deployment of ipv6 server is possible and nothing else seems to be broken.

**Comment by jyejare** [comment]
The sanity regression testing is successfully verified for Ipv4 after recent changes. An Ipv6 was tested previously and it was successful too.

CC: @JacobCallahan 

@devendra104 , I think we need to merge the changes in CI for this PR to me merged. Especiallly conf related changes which wont also impact existing IPv4 pipelines.

**Comment by jyejare** [comment]
trigger: test-robottelo
pytest: tests/foreman -m build_sanity

**Comment by jyejare** [comment]
trigger: test-robottelo
pytest: tests/foreman -m build_sanity

**Comment by jyejare** [comment]
trigger: test-robottelo
pytest: tests/foreman -m build_sanity

File: robottelo/utils/url.py
Comment (Griffin-Sullivan) [reply robottelo/utils/url.py:21]: Sorry I should check my direct mentions more often.

So to clarify this is a problem with the fam.yaml settings because it uses this notation?

```yaml
FAM:
...
  COMPUTE_PROFILE:
      libvirt:
        ...
          compute_profile:
            ...
            attrs:
              ...
              nics_attributes:
                0:
                  ...
```

The reason I have it like this is because that's how https://github.com/theforeman/foreman-ansible-modules/blob/develop/tests/test_playbooks/vars/compute_profile.yml is set up. If we want to change the notation we just need to update FAM as well. Let me know if I need to make the changes.

**Comment by jyejare** [comment]
```
trigger: test-robottelo
pytest: tests/foreman -m build_sanity
env:
    ROBOTTELO_server__deploy_arguments__deploy_network_type: ipv4
    ROBOTTELO_server__is_ipv6: False
    ROBOTTELO_server__http_proxy_ipv6_url: ''
```

**Comment by Satellite-QE** [comment]
**PRT Result**
```
Build Number: 7388
Build Status: SUCCESS
PRT Comment: pytest tests/foreman -m build_sanity --external-logging
Test Result : ======= 12 passed, 5502 deselected, 5607 warnings in 1555.35s (0:25:55) ========
```

**Comment by jyejare** [comment]
@JacobCallahan We should be good to merge the PR now as the PRT is passing `showing no regression on existing IPv4 testing`.

This should not need CI MR to merged as we are taking care of it from dynaconf validators.

**Comment by JacobCallahan** [comment]
I'm going to merge this now. @ogajduse if there is anything you'd still like addressed for this post-merge, please write it down and create an issue for it to be resolved later. I'd like to move this forward so we can flesh out things live.

File: robottelo/utils/url.py
Comment (jyejare) [reply robottelo/utils/url.py:21]: @Griffin-Sullivan Yes, Good to have uniformed behavior across robottelo.

**Pull request merged**
